esp/platform/espcache.cpp

```diff
@@ -0,0 +1,302 @@
+#endif
+}
+
+ESPMemCached::~ESPMemCached()
+{
+    if (pool)
+    {
+        memcached_pool_release(pool, connection);
+        connection = nullptr;//For safety (from changing this destructor) as not implicit in either the above or below.
+        memcached_st *memc = memcached_pool_destroy(pool);
+        if (memc)
+            memcached_free(memc);
+    }
+    else if (connection)//This should never be needed but just in case.
+    {
+        memcached_free(connection);
+    }
+}
+
+bool ESPMemCached::init(const char * _options)
+{
+    CriticalBlock block(cacheCrit);
+
+    if (initialized)
+        return initialized;
+
+    if (!isEmptyString(_options))
+        options.set(_options);
+    else
+        options.set("--SERVER=127.0.0.1");
+    pool = memcached_pool(options.get(), options.length());
+    assertPool();
+
+    setPoolSettings();
+    connect();
+    if (connection)
+        initialized = checkServersUp();
+    return initialized;
+}
+
+bool ESPMemCached::cacheResponse(const char* cacheID, unsigned cacheSeconds, const char* content, const char* contentType)
+{
+    VStringBuffer contentTypeID("ContentType_%s", cacheID);
+
+    CriticalBlock block(cacheCrit);
+    ESP_CACHE_RETURN ret = set("ESPResponse", cacheID, content, cacheSeconds);
+    if (ret != ESP_CACHE_SUCCESS)
+        return false;
+    return set("ESPResponse", contentTypeID.str(), contentType, cacheSeconds) == ESP_CACHE_SUCCESS;
+}
+
+bool ESPMemCached::readResponseCache(const char* cacheID, StringBuffer& content, StringBuffer& contentType)
+{
+    VStringBuffer contentTypeID("ContentType_%s", cacheID);
+
+    CriticalBlock block(cacheCrit);
+    if (!checkAndGet("ESPResponse", cacheID, content))
+        return false;
+    return checkAndGet("ESPResponse", contentTypeID.str(), contentType);
+}
+
+void ESPMemCached::setPoolSettings()
+{
+    assertPool();
+    const char * msg = "memcached_pool_behavior_set failed - ";
+    assertOnError(memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_KETAMA, 1), msg);//NOTE: alias of MEMCACHED_DISTRIBUTION_CONSISTENT_KETAMA amongst others.
+    memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_USE_UDP, 0);  // Note that this fails on early versions of libmemcached, so ignore result
+    assertOnError(memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_NO_BLOCK, 0), msg);
+    assertOnError(memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_CONNECT_TIMEOUT, 1000), msg);//units of ms.
+    assertOnError(memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_SND_TIMEOUT, 1000000), msg);//units of mu-s.
+    assertOnError(memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_RCV_TIMEOUT, 1000000), msg);//units of mu-s.
+    assertOnError(memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_BUFFER_REQUESTS, 0), msg);
+    assertOnError(memcached_pool_behavior_set(pool, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL, 1), "memcached_pool_behavior_set failed - ");
+}
+
+void ESPMemCached::connect()
+{
+    assertPool();
+    if (connection)
```

Comment:
This works, but I would find it more readable if the "if (connection)" were duplicated in each #if below.  It groups the logic closer together.